### PR TITLE
[PM-11649] Expired link services

### PR DESCRIPTION
**Pull request by andrebispo5**
## 🎟️ Tracking

[PM-11649](https://bitwarden.atlassian.net/browse/PM-11649) Email verification expired token service

## 📔 Objective

Consume api endpoint to verify if the token provided by email, during email verification flow, is still valid to use.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-11649]: https://bitwarden.atlassian.net/browse/PM-11649?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt [hunk 1]
```diff
@@ -127,4 +129,22 @@ class IdentityServiceImpl(
...
+        body: VerifyEmailTokenRequestJson,
+    ): Result<VerifyEmailTokenResponseJson> {
+        return unauthenticatedIdentityApi
+            .verifyEmailToken(body = body)
+            .map { VerifyEmailTokenResponseJson.Success }
+            .recoverCatching { throwable ->
+                val bitwardenError = throwable.toBitwardenError()
+                bitwardenError
+                    .parseErrorBodyOrNull<VerifyEmailTokenResponseJson.Invalid>(
+                        codes = (400..499).toList(),
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/network/service/IdentityServiceImpl.kt:144]: Do we actually want to catch all of the 400s?

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt [hunk 2]
```diff
@@ -1256,6 +1258,32 @@ class AuthRepositoryImpl(
...
+                onSuccess = {
+                    VerifyEmailTokenResult.Verified
+                },
+                onFailure = {
+                    // Server sends a message if the link is expired, this is the only way
+                    // to check if the token is expired
+                    if (it.message?.contains("Expired link", ignoreCase = true) == true) {
+                        return VerifyEmailTokenResult.LinkExpired
+                    } else {
+                        return VerifyEmailTokenResult.Error
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1282]: You do not need these `return`s

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt [hunk 3]
```diff
@@ -1256,6 +1258,32 @@ class AuthRepositoryImpl(
...
+                ),
+            )
+            .fold(
+                onSuccess = {
+                    VerifyEmailTokenResult.Verified
+                },
+                onFailure = {
+                    // Server sends a message if the link is expired, this is the only way
+                    // to check if the token is expired
+                    if (it.message?.contains("Expired link", ignoreCase = true) == true) {
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1279]: I just spoke to @dseverns-livefront about this. Stuff like this should be in the service.

**Review by dseverns-livefront — CHANGES_REQUESTED** [review]
Looks like some Jira wires got crossed: I think #3862 may be taking care of this, the main difference I am seeing is the wider range of 400s errors caught. Lets chat tomorrow and figure it out! 👍

**Comment by andrebispo5** [comment]
>  [ 68747470733a2f2f6d65646961302e67697068792e636f6d2f6d656469612f6c33366b553830785066306f6a47304572672f67697068792e676966 ](https://camo.githubusercontent.com/0d4060d90d4e82937f382c171bd541be46acb0862d7036b7d14749ef73e4e4d8/68747470733a2f2f6d65646961302e67697068792e636f6d2f6d656469612f6c33366b553830785066306f6a47304572672f67697068792e676966) [ ](https://camo.githubusercontent.com/0d4060d90d4e82937f382c171bd541be46acb0862d7036b7d14749ef73e4e4d8/68747470733a2f2f6d65646961302e67697068792e636f6d2f6d656469612f6c33366b553830785066306f6a47304572672f67697068792e676966)
> 
> Looks like some Jira wires got crossed: I think #3862 may be taking care of this, the main difference I am seeing is the wider range of 400s errors caught. Lets chat tomorrow and figure it out! 👍

Yup 

>  [ 68747470733a2f2f6d65646961302e67697068792e636f6d2f6d656469612f6c33366b553830785066306f6a47304572672f67697068792e676966 ](https://camo.githubusercontent.com/0d4060d90d4e82937f382c171bd541be46acb0862d7036b7d14749ef73e4e4d8/68747470733a2f2f6d65646961302e67697068792e636f6d2f6d656469612f6c33366b553830785066306f6a47304572672f67697068792e676966) [ ](https://camo.githubusercontent.com/0d4060d90d4e82937f382c171bd541be46acb0862d7036b7d14749ef73e4e4d8/68747470733a2f2f6d65646961302e67697068792e636f6d2f6d656469612f6c33366b553830785066306f6a47304572672f67697068792e676966)
> 
> Looks like some Jira wires got crossed: I think #3862 may be taking care of this, the main difference I am seeing is the wider range of 400s errors caught. Lets chat tomorrow and figure it out! 👍

Yup it is the same... Just added the wider range because it was like that on iOS but I don't think it is necessary. I'll close and link to #3862.

**Pull request closed without merging**